### PR TITLE
fix(access-control): [5/6] provider-agnostic security/correctness fixes

### PR DESCRIPTION
**Pull request by amitgilad3**
## Summary
Part 5 of 6, split out of #234. Stacked on #252.

Hand-split from a commit in #234 that bundled these fixes together with the new Cerbos provider (kept separate — see #TBD, part 6). Also includes the `examples/with-opa-oidc/` demo (real OIDC identity flow that several of these fixes touch).

Fixes surfaced by a full review of the OPA access-control work:
- A genuine SQL parse failure in `extract_resources()` was indistinguishable from "no tables referenced," silently bypassing `onMissingSchema: deny`.
- A partial OPA response (missing a verdict for one of several requested resources) was treated as allowed for that resource instead of denied.
- Cached query results bypassed access control entirely on replay — a revoked or tightened policy had no effect until the cache entry expired.
- The read/write invariant check only ran for the subset of queries that got row-filtered/masked; `run_fixup_scripts` now rejects any statement-kind change unconditionally.
- The external guard webhook/script payload had silently dropped the pre-translation SQL field; restored via a new `GuardContext.original_sql` field.
- `StaticCatalogProvider` (and the e2e `MapCatalog` test double) ignored catalog/database qualifiers, letting same-named tables in different catalogs collide.
- MCP sessions without a declared dialect skipped schema resolution entirely.
- An OAuth2 client-credentials token refresh released its lock between check and fetch, causing a thundering herd of duplicate token requests.
- `AccessController`'s decision cache switched from a single global `Mutex<HashMap>` to a sharded `DashMap` on the per-query hot path.
- Removed unused `GuardChainOutcome::Proceed{sql}` generality.
- Deduplicated `OpaAccessGuard`'s reimplementation of `AccessControlConfig`'s own group-resolution logic, and the deny-and-record blocks in `dispatch.rs`.
- Surfaced the response body on non-2xx provider errors (previously discarded).

## Testing
`cargo check --workspace` passes on this branch (verified before pushing). Existing unit/e2e suites cover the fixed paths.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/queryflux-access-control/src/providers/opa/mod.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            .ok()?;
+        let body: serde_json::Value = resp.json().await.ok()?;
+        let token = body.get("access_token")?.as_str()?.to_string();
+        let expires_in = body
+            .get("expires_in")
+            .and_then(|v| v.as_u64())
+            .unwrap_or(300);
+        let exp = Instant::now() + Duration::from_secs(expires_in);
+        *guard = Some((token.clone(), exp));
+        Some(token)
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-access-control/src/providers/opa/mod.rs]: Fixed in 8f287bf2. Both OPA and the OpenFGA client's token POST now use an explicit timeout, and a failed fetch is cached as a short cooldown (5s) — a caller queued behind the refresh lock during an IdP outage now returns `None` immediately during that cooldown instead of repeating the same failing request, so total latency no longer grows linearly with the number of waiters.

File: crates/queryflux-access-control/src/providers/opa/mod.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            } else {
+                PolicyError::Transport(e.to_string())
+            }
+        })?;
+
+        let status = resp.status();
+        if !status.is_success() {
+            let body = resp.text().await.unwrap_or_default();
+            return Err(PolicyError::Status(status.as_u16(), body));
+        }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-access-control/src/providers/opa/mod.rs]: Fixed in 8f287bf2. Added `truncate_error_body` (UTF-8-boundary-safe) and applied it before constructing `PolicyError::Status`, matching the doc comment's "best-effort, truncated" claim. Unit tests cover short bodies, long bodies, and a multi-byte cut point.

File: crates/queryflux-access-control/src/providers/opa/wire.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        if !decided.contains(&req.table) {
+            resources.push(ResourceDecision {
+                table: req.table.clone(),
+                allow: false,
+                reason: Some("policy returned no decision for this resource".to_string()),
+                row_filters: Vec::new(),
+                column_masks: Vec::new(),
+            });
+        }
+    }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-access-control/src/providers/opa/wire.rs]: Fixed in 8f287bf2. `from_response` now produces one `ResourceDecision` per *requested* resource (not per response entry), with `table` always normalized to the request's own `qualified_name()`. A bare policy reply only resolves a qualified request when that bare name is unique across the whole request — otherwise it's treated as undecided (denied) rather than guessed at, so a single bare `orders` reply can no longer cover both `sales.orders` and `hr.orders`. Added regression tests for the ambiguous case, the disambiguated (qualified) case, and that the output table is always the canonical qualified name. I also checked the translation layer's own scan-site matching (`find_policy` in access.rs) — it already fails closed on the same kind of ambiguity, so this fix keeps both layers consistent.

File: crates/queryflux-e2e-tests/tests/access_control_rewrite_tests.rs [hunk 4]
```diff
@@ -0,0 +1,1106 @@
...
+        .await
+        .expect("select");
+    // Named REDACT renders as regexp_replace without a global flag; Postgres/DuckDB
+    // then replace only the first alphanumeric. The result must still not be the raw SSN.
+    assert_ne!(rows[0][0], "111-22-3333");
+    assert!(
+        rows[0][0].contains('x'),
+        "REDACT must replace at least one alphanumeric with x, got {:?}",
+        rows[0][0]
+    );
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-e2e-tests/tests/access_control_rewrite_tests.rs]: Fixed in 8f287bf2, with your suggested dialect split. `Postgres`/`DuckDb` now render with the `g` flag; `Trino`/`Athena` are left alone since their 3-arg form already replaces every match and a 4th positional argument there is a capture-group index, not a flag. Strengthened both the translation-layer unit test (asserts the exact rendered SQL per dialect) and the e2e test (asserts the SSN is fully redacted to `xxx-xx-xxxx`, not just partially). I scoped this to the two dialects I could verify rather than guessing at flag syntax for the rest (some, like Trino/Snowflake/MySQL, already default to global-replace, where adding `g` would be wrong) — left a comment noting that.

File: crates/queryflux-frontend/src/access_control_guard.rs [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+        {
+            Ok(r) => r,
+            Err(e) => {
+                return match conn.on_missing_schema {
+                    OnMissingSchema::Deny => GuardResult::deny(
+                        format!("access control: could not analyze query: {e}"),
+                        "ACCESS_ANALYSIS_FAILED",
+                    ),
+                    OnMissingSchema::Evaluate => GuardResult::allow(),
+                }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-frontend/src/access_control_guard.rs]: This is already fixed as of the earlier #251/#252 review rounds — `check()`'s `Err` branch has unconditionally denied (no `onMissingSchema`-gated allow) since then, and that's what's on this branch post-rebase. This comment predates the rebase (it's anchored to the old pre-rebase commit 3b047459); reconfirmed against 8f287bf2 rather than re-touched.

File: crates/queryflux-frontend/src/access_control_guard.rs [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+/// `CREATE TABLE … AS`, …) are still policy-checked — see `check`.
+///
+/// When `polyglot-sql` couldn't parse the statement at all, fall back to the same string
+/// heuristic the built-in guards use (`is_read_like_fallback`) rather than defaulting to
+/// "skip the stage" — a query that heuristically looks like a read should still be
+/// evaluated even if this particular parser choked on it (resource extraction uses a
+/// different, Python-sqlglot-based parser, so a `polyglot-sql` failure here doesn't imply
+/// extraction will also fail).
+fn classify_operation(stmts: Option<&[Expression]>, sql: &str) -> Operation {
+    match stmts.and_then(|s| s.first()) {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-frontend/src/access_control_guard.rs]: Fixed in 8f287bf2, per your suggested diff. The guard now denies a batch with more than one parsed statement before `classify_operation`/`extract_resources` ever run, since both only look at the first statement. Added an e2e test that sends `SELECT 1; DELETE FROM orders` through the real Postgres simple-query path (the one that doesn't split multi-statement batches) with a row filter configured, and verifies both that it's rejected and that the DELETE never reached the engine (row count unchanged).

File: crates/queryflux-frontend/src/admin.rs [hunk 7]
```diff
@@ -2725,6 +2745,341 @@ async fn put_guardrails_config_handler(
...
+        };
+        let incoming_token_empty = opa
+            .get("bearerToken")
+            .and_then(|t| t.as_str())
+            .is_none_or(|s| s.is_empty());
+        if incoming_token_empty {
+            if let Some(prev) = prev_opa.get("bearerToken") {
+                opa.insert("bearerToken".into(), prev.clone());
+            }
+        }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-frontend/src/admin.rs]: Fixed in 8f287bf2. Added a `same_endpoint` check so a retained `bearerToken`/`clientSecret` is only kept when `opa.url`/`clientCredentials.tokenEndpoint` is unchanged from the previous config. Added a regression test for the exact scenario you described (blank the secret while redirecting the endpoint elsewhere) and fixed the two existing tests' fixtures, which were missing a matching endpoint and so didn't exercise this path at all.

File: crates/queryflux-guardrails/src/built_in.rs [hunk 8]
```diff
@@ -314,43 +306,61 @@ fn simple_glob_match(haystack: &str, pattern: &str) -> bool {
...
+        roles: Vec<String>,
+        attributes: BTreeMap<String, serde_json::Value>,
+        session_extra: HashMap<String, String>,
     }
 
     impl TestCtx {
-        fn new(sql: &str, translated: &str) -> Self {
+        /// `translated` is accepted for call-site compatibility; guards now run on the
+        /// source SQL, so `sql` is what the context carries.
+        fn new(sql: &str, _translated: &str) -> Self {
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-guardrails/src/built_in.rs]: Fixed in 8f287bf2, per your suggested diff. Dropped the unused second parameter and collapsed all 48 call sites; fixed the one that relied on it (`read_only_allows_leading_whitespace_select` now actually exercises the whitespace-prefixed SELECT).

File: crates/queryflux-translation/src/sqlglot.rs [hunk 9]
```diff
@@ -357,23 +372,96 @@ fn run_fixup_scripts(
...
+                    "translation script {i} transform() call failed: {e}"
+                ))
+            })?;
+
+        current = result.extract().map_err(|e| {
             QueryFluxError::Translation(format!(
-                "translation script {i} transform() call failed: {e}"
+                "translation script {i} transform() must return a str: {e}"
             ))
         })?;
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-translation/src/sqlglot.rs]: Addressed in 8f287bf2, via startup/reload validation rather than a compatibility adapter (translating an in-place AST-mutation contract into string-in/string-out isn't something I could do safely/generically). Added `queryflux_translation::sqlglot::validate_fixup_script`, which smoke-tests a script's `transform` against a trivial probe query. YAML `translation.pythonScripts` entries now abort startup on failure (consistent with this codebase's fail-startup-on-bad-explicit-config convention); DB-persisted per-group scripts are filtered out (with a loud `tracing::error!` naming the group and index) at both startup and reload, since those are Studio-editable and one stale script shouldn't block the whole proxy from starting. Unit tests cover both the validator itself and the per-group filtering.

File: crates/queryflux/src/main.rs [hunk 10]
```diff
@@ -956,6 +957,11 @@ async fn main() -> Result<()> {
     let guard_script_bodies =
         load_guard_script_bodies(backend.as_deref().map(|b| b as &dyn AdminStore)).await;
 
+    // --- Build the data access-control guard (OPA row filtering / column masking /
+    // table-column allow-deny), if configured. Runs on the source SQL, before
+    // dialect translation — see `queryflux_frontend::access_control_guard`.
+    let access_control_guard = build_access_control_guard(&config);
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux/src/main.rs]: Already fixed as of the #252 review round — startup now reads a persisted `access_control_config` and applies it over YAML before `LiveConfig` is built, aborting startup if it can't be read or built. This comment is anchored to the old pre-rebase commit; reconfirmed against 8f287bf2.

File: examples/with-opa-oidc/README.md [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+docker compose --profile seed run --rm data-seed
+# UI: http://127.0.0.1:8183  (Ask OPA works immediately)
+
+# 2. Python deps used by translation / scan-site rewrite
+cd ../..
+queryflux --install-deps   # or: python3 -m venv .venv && .venv/bin/pip install -r requirements.txt
+
+# 3. QueryFlux from this branch — verifies Keycloak JWTs, Postgres on :5434,
+#    catalogProvider at Lakekeeper :8181
+cargo run -p queryflux -- --config examples/with-opa-oidc/config.yaml
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply examples/with-opa-oidc/README.md]: Fixed in 8f287bf2 — pointed both commands at `config-no-catalog.yaml`, the file that actually exists in this example.

File: website/docs/access-control/customer-api-row-filters.md [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+```yaml
+auth:
+  provider: static
+  required: true
+  static:
+    users:
+      - username: api-service-a
+        # bcrypt hash of the service password
+        passwordHash: "..."
+        groups: [customer_portal]
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply website/docs/access-control/customer-api-row-filters.md]: Fixed in 8f287bf2, with your suggested schema (`staticUsers.users` map keyed by username, matching `examples/with-opa/config.yaml`).

File: website/docs/architecture/guardrails.md [hunk 13]
```diff
@@ -33,6 +34,20 @@ This lets you apply baseline safety globally (e.g. read-only for all agents) whi
...
+
+| Config | Guard name | When it runs | Studio |
+| --- | --- | --- | --- |
+| `accessControl:` | `opa_access` | Before dialect translation | Guard Actions (`rewrite` / `deny`) + **Rewritten SQL (access control)** panel |
+
+Access-control **policy** (Rego, filters, masks, dry-run) is documented under **[Access control](../access-control/overview)** — provider docs: **[OPA](../access-control/opa)**. Observability is the same `guard_actions` array as every other guard.
+
+When access control is on, the rewriting guard is `opa_access`.
+
+**Scope** mirrors guardrails' global + `groups` pattern, but lives under `accessControl:` (edited on Studio **Access Control**, not **Guardrails** or **Clusters**): global `enabled` plus optional `groups.<clusterGroup>.enabled` to skip OPA for sandbox groups or opt in only where needed. One OPA connection.
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply website/docs/architecture/guardrails.md]: Already fixed as of the #252 review round — the scope paragraph now describes named `connections`, `defaultConnection`, and `groups.<clusterGroup>.connection` instead of "One OPA connection." Reconfirmed against 8f287bf2.

File: website/docs/architecture/guardrails.md [hunk 14]
```diff
@@ -120,39 +135,45 @@ Queries routed to the `analysts` group run: `read_only` → `row_limit(100000)`.
...
 
 ```yaml
 guardrails:
   global:
     - kind: http_webhook
       url: "https://hooks.example.com/guard"
-      timeout_ms: 5000
-      fail_behavior: deny   # or: allow
+      timeout_ms: 5000   # default 1000, capped at 30000
+      retry_count: 2     # retries on 5xx only; default 0
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply website/docs/architecture/guardrails.md]: Fixed in 8f287bf2, per your suggested diff.

File: crates/queryflux-access-control/src/controller.rs [hunk 15]
```diff
@@ -196,94 +188,53 @@ impl AccessController {
...
-    let mut resources: Vec<ResourceKey> = req
+fn cache_key(req: &AccessRequest) -> u64 {
+    let mut h = std::collections::hash_map::DefaultHasher::new();
+    req.identity.user.hash(&mut h);
+    hash_sorted(&mut h, req.identity.groups.iter().map(String::as_str));
+    hash_sorted(&mut h, req.identity.roles.iter().map(String::as_str));
+    for (k, v) in &req.identity.attributes {
+        k.hash(&mut h);
+        v.to_string().hash(&mut h);
+    }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-access-control/src/controller.rs]: Fixed in 8f287bf2. Replaced the digest-only `DashMap<u64, _>` with a typed `CacheKey` (`Hash + Eq`) built from structured fields rather than delimiter-joined strings. Added 5 regression tests, including the exact collision you identified (groups=["a","b"] roles=[] vs groups=["a"] roles=["b"]) and the delimiter-collision cases for resource identifiers and column names — all of which fail against the old digest and pass against the new key.

File: crates/queryflux-frontend/src/dispatch.rs [hunk 16]
```diff
@@ -2279,26 +2266,81 @@ pub async fn execute_to_sink(
...
+                    &session,
+                    Some(&schema_context),
+                    &effective_tags,
+                )
+                .await
+                {
+                    AccessStageOutcome::Denied { reason, .. } => {
+                        return sink.on_error(&reason).await
+                    }
+                    AccessStageOutcome::Rewritten { .. } => skip_cache_lookup = true,
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-frontend/src/dispatch.rs]: Fixed in 8f287bf2, per your suggested approach: the rewrite flag is now tracked outside the `cache_key` block and the cache-write path skips creating a `cache_writer` when access control rewrote the query. I verified this by inspection and the existing dispatch test suite (all pass); I didn't add a new e2e test since none of the e2e harness constructors currently wire in a real result cache (they all use `NoopResultCache`) — building that infra felt disproportionate to a Minor finding, but flagging it in case you'd weigh that differently.

File: website/docs/access-control/overview.md [hunk 17]
```diff
@@ -208,6 +208,16 @@ Studio's **Access Control** page can add, edit, and remove any number of named c
 
 ---
 
+## What policies apply to
+
+Policies govern **reads**. Any table a statement *reads* is evaluated as `table.select` (deny, row filter, column mask) — including reads inside a write: `INSERT … SELECT`, `CREATE TABLE … AS SELECT`, `CREATE VIEW … AS SELECT`, `MERGE … USING`, and subqueries in `UPDATE`/`DELETE`. A protected table cannot be copied out through a write statement.
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply website/docs/access-control/overview.md]: Fixed in 8f287bf2, per your suggested framing. Also revised the adjacent "CREATE… not evaluated at all" sentence: that's true for the CREATE/DROP/ALTER target itself, but CTAS/CREATE VIEW AS SELECT's *source* SELECT is still evaluated as a read when table.select is enabled — the sentence previously read as if the whole statement were skipped.

File: crates/queryflux-access-control/src/providers/opa/mod.rs [hunk 18]
```diff
@@ -120,3 +159,47 @@ impl PolicyDecisionProvider for OpaProvider {
...
+    }
+
+    /// A multi-byte character must never land the cut mid-codepoint (`String::truncate`
+    /// panics on that), even when the limit itself falls inside one.
+    #[test]
+    fn truncate_error_body_respects_utf8_boundaries() {
+        let long = "é".repeat(ERROR_BODY_LIMIT); // each "é" is 2 bytes, so this crosses the limit
+        let out = truncate_error_body(long);
+        assert!(out.ends_with("... (truncated)"));
+    }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-access-control/src/providers/opa/mod.rs]: Fixed in b5b4b119, per your suggested diff — added the one-byte prefix and asserted `!is_char_boundary(ERROR_BODY_LIMIT)` directly so the test can't silently stop exercising the boundary-seeking loop again.

File: crates/queryflux-auth/src/authorization.rs [hunk 19]
```diff
@@ -275,41 +290,63 @@ impl OpenFgaAuthorizationClient {
...
+                    Some((token, ttl)) => {
+                        *guard = Some(OpenFgaTokenCacheEntry::Valid(token.clone(), now + ttl));
+                        Some(token)
+                    }
+                    None => {
+                        *guard = Some(OpenFgaTokenCacheEntry::FailedUntil(
+                            now + OPENFGA_TOKEN_FAILURE_COOLDOWN,
+                        ));
+                        None
+                    }
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-auth/src/authorization.rs]: Fixed in fa1d6608 for both sites, per your suggested diff — both now take a fresh `Instant::now()` right before storing `FailedUntil` instead of reusing the pre-fetch `now`.

File: crates/queryflux-frontend/src/dispatch.rs [hunk 20]
```diff
@@ -2279,26 +2273,84 @@ pub async fn execute_to_sink(
...
+                    &effective_tags,
+                )
+                .await
+                {
+                    AccessStageOutcome::Denied { reason, .. } => {
+                        return sink.on_error(&reason).await
+                    }
+                    AccessStageOutcome::Rewritten { .. } => {
+                        skip_cache_lookup = true;
+                        access_control_rewrote_for_cache = true;
```
[the remark this answers is not in the document]
Comment (amitgilad3) [reply crates/queryflux-frontend/src/dispatch.rs]: Fixed in fa1d6608. Since the real engine genuinely isn't resolvable at this point (routing/cluster-slot acquisition happens later, deliberately, so a cache hit can skip it), I went with your second option: `Denied` under the synthetic `EngineType::Cache` now falls through like `Rewritten` already did — skip cache, proceed to `execute_to_sink_inner`, which reaches `setup_sync_query`'s real per-engine access-control check. A genuinely-denied query is still denied there for real; this only stops a synthetic-engine artifact from being the final word.
